airflow/models/dagrun.py

```diff
@@ -286,7 +287,17 @@ def find(
     @staticmethod
     def generate_run_id(run_type: DagRunType, execution_date: datetime) -> str:
         """Generate Run ID based on Run Type and Execution Date"""
-        return f"{run_type}__{execution_date.isoformat()}"
+        from airflow.configuration import conf
+        local_run_id = conf.getboolean("core", "localize_dag_run_id", fallback=False)
+        if local_run_id:
+            local_time = pendulum.instance(execution_date).astimezone(tz=settings.TIMEZONE)
```

Comment:
It should be based on the _dag_ timezone, not the local timezone.

Reply:
it makes sense. 
But we have to change the method from static_method to class method.
I will change it tomorrow

Reply:
Oh hmmm. It can't even be a class method. Maybe we should just pass the date in the "right" TZ here already?

Reply:
good idea. It leads to a small change for the project.

Reply:
@ashb 
I thought we can get the dag timezone by the `start_date` or `end_date`, but actually, the two values may be `None if the user doesn't set them.
So the question is that which `timezone` do you mean? 
The timezone of `execution_date`?

Reply:
Just confirm that the `dag timezone` is the value that is inited here? 
https://github.com/apache/airflow/blob/e99624d2f17f0aadcb992827e9e721f5e6c978d8/airflow/models/dag.py#L326

Reply:
@ashb 
I checked the codes, we can not use the `tzinfo` in the execution_date directly, because it has been transformed into UTC already.
![image](https://user-images.githubusercontent.com/7628879/129824951-cf0ff5bd-916e-42af-82eb-1c9315afbeab.png)

So we have two ways to implement:
1. change the method generate_run_id from static method to class method
2. add a parameter `timezone` in the method `generate_run_id` and the default value is settings.TIMEZONE.

Which one do you prefer? Or there is something I miss that we can have a better choice?
Let me know what do you think.
Thanks